semgrep-core/src/cli/Main.ml

```diff
@@ -402,7 +402,23 @@ let cache_computation file cache_file_of_file f =
           res)
         else
           let res = f () in
-          Common2.write_value (Version.version, file, res) file_cache;
+          let cleanup_cache () =
+            if Sys.file_exists file_cache then Sys.remove file_cache
+          in
+          Sys.set_signal Sys.sigxfsz
+            (Sys.Signal_handle
+               (fun _ ->
+                 logger#error
+                   "Could not write cache file for %s: file size limit exceeded"
+                   file;
+                 cleanup_cache ()));
```

Comment:
I think this signal handler is unsafe because it's going to execute asynchronously, similar to a thread. If you want to write a message to stderr, use `Unix.write` (unbuffered write) instead of the buffered write that `logger#error` presumably uses. In other words, you probably don't want to do that at all. The simplest thing to do is to have the signal handler raise an exception, and have this exception be caught in the normal execution flow (main thread).

Also, registering this signal handler should probably be done once and for all during program initialization (wherever we do these things).

e.g.
```
exception File_size_limit_exceeded
let () = Sys.signal_handler (Sys.Signal_handle (fun _signal -> raise File_size_limit_exceeded)
```
and then catch `File_size_limit_exceeded` where you think it's relevant.

Reply:
Great feedback, I'll fix that.

Reply:
If we raise an exception inside the handler this can pop up anywhere and it's very useful to recover from it. It seems that if you simply ignore it, that will cause the sys call to fail, and you get a `Sys_error` instead, from which it is easy to catch and recover from. Let me know what do you think!

Reply:
If that's how it works, let's try this.